cmd/xl-storage-format-v2.go

```diff
@@ -241,14 +241,65 @@ type xlMetaV2 struct {
 // tempXlMetaV2 returns a temporary xlMetaV2.
 // When usage is done, return control via the returned function.
 func tempXlMetaV2() (x *xlMetaV2, reUse func()) {
-	return &xlMetaV2{}, func() {}
+	x = xlMetaV2Pool.Get().(*xlMetaV2)
+	x.clear()
```

Comment:
@klauspost isn't it simple to do 

```
x = xlMetaV2Pool.Get().(*xlMetaV2)
return x, func() {
          x = &xlMetaV2{}
          xlMetaV2Pool.Put(x)
}
```

shouldn't this handle all the clearing complications?  would this cause more allocations? 

It looks like it has become very tricky to un-initialize each layers and its looking trickier if hit a bug.  Are you sure that this PR would fix something significant?

Reply:
@harshavardhana The problem isn't with xlMetaV2 itself, but the slices/maps deeply nested within it. Those are what is taking up the majority of allocations.

The problem is that when they have `omitempty` they are *not* cleared by unmarshal when they have content.

Take this:
```
type xlMetaV2Version struct {
	Type         VersionType           `json:"Type" msg:"Type"`
	ObjectV1     *xlMetaV1Object       `json:"V1Obj,omitempty" msg:"V1Obj,omitempty"`
	ObjectV2     *xlMetaV2Object       `json:"V2Obj,omitempty" msg:"V2Obj,omitempty"`
	DeleteMarker *xlMetaV2DeleteMarker `json:"DelObj,omitempty" msg:"DelObj,omitempty"`
}
```
When a version had previously been a Delete Marker it would have a non-nil `DeleteMarker`. This is nice if the new one is also a delete marker, because it can be reused and we avoid an alloc.

However, if the new object has `ObjectV2` the `DeleteMarker` is never cleared because it is omitted from the serialized data and the unmarshaler therefore leaves it alone. I tried cleaning them all up, but there still appears to be problems, so I have now removed the pool, since this is high risk.

Reply:
(also I presume you mean `*x = xlMetaV2{}` to clear it and not alloc a new.